### DISCUSSION TITLE
Impact of changes React is making to <form>

### DISCUSSION THREAD

**Discussion by petejodo** [ariakit/ariakit#3744] [General]
With the release of [React 19 beta](https://react.dev/blog/2024/04/25/react-19), there seems to be quite a lot of work done towards forms. I'm curious if any ideas, thoughts, opinions, feelings, etc have been given towards how forms in Ariakit work to better align with these new React features.

**Comment by diegohaz** [comment]
We should definitely update Ariakit Form to align with the new React patterns, but I haven't given it much consideration yet. I'm open to suggestions.

**Reply by petejodo** [reply]
I haven't given much consideration myself but knowing you're receptive to it, I'll gather thoughts for when I find myself working w/ Forms which should be quite a bit in the near future. Thanks!

**Comment by petejodo** [comment]
I do find it easier to just use the base `<form>` than ariakit's `Form`/`useFormStore` via the `action` prop even though you lose types and have to deal w/ `FormData` e.g.

```tsx
function submit(formData: FormData) {
  // ...
}

return <form action={submit}>...</form>
```

The only difficulty with that is that some ariakit components don't set a `value` that gets picked up by the `<form>` e.g. combobox and therefore requires a fair amount of work to build html `<form>`-friendly components (albeit there are examples for how to do that). This would then naturally work with all the `useActionState` hooks. I guess ariakit's form store could also implement `action` somehow 🤔 e.g.

```tsx
const form = useFormStore(...);

const formAction = form.useSubmit(() => {
  // can be server action
});

return <form action={formAction}>...</form>

// or maybe still have ariakit's own Form...
return <Form action={formAction}>...</Form>
```

**Comment by petejodo** [comment]
So I have an example of some functionality that I recently implemented which I used `useActionState` w/ a server action hooked up to a Combobox/MenuList. I had to use a `MenuList` for the following to work because I could not get this working with `ComboboxItem` where I could w/ `MenuItem`. This is bit of a reconstructed example so it's more for demonstration purposes rather than expecting it to work. Basically it's a combobox to search for a company which the user can then select to set that company to an arbitrary record

```tsx
const formRef = useRef<HTMLFormElement>(null);
const [companies, searchCompaniesAction, isPending] = useActionState(searchCompanies, []);
const [, selectCompanyAction] = useActionState(selectCompany, undefined);

let menuItems: React.ReactNode;

if (isPending) {
  menuItems = <div>Searching...</div>
} else if (companies.length === 0) {
  menuItems = <div>No Results</div>
} else {
  menuItems = companies.map((company) => (
    <MenuItem
      key={company.id}
      hideOnClick
      render={
        <button
          formAction={selectCompanyAction}
          type="submit"
          name="company"
          value={company.id}
        />
      }
    >
      {company.name}
    </MenuItem>
  ));
}

return (
  <form ref={formRef} action={searchCompaniesAction}>
    <input type="hidden" name="record" value={record.id} />

    <ComboboxProvider resetValueOnHide setValue={() => formRef.current?.requestSubmit()}>
      <MenuProvider>
        <Combobox name="query" />
        <ComboboxPopover>
          <MenuList alwaysOpen>
            {menuItems}
          </MenuList>
        </ComboboxPopover>
      </MenuProvider>
    </ComboboxProvider>
  </form>
)
```

**Comment by petejodo** [comment]
another thought is the increasing usage of zod and similar libs so an API that lines up nicely with them I think is important. Currently ariakit forms don't support form-level errors which would be necessary to play nice with those libs
